tensorboard/uploader/proto/write_service.proto

```diff
@@ -119,6 +121,82 @@ message WriteScalarResponse {
   // This is empty on purpose.
 }
 
+// The blob writing procedure works as follows:
+//  * The caller instantiates a WriteBlob RPC, which is a bidirectional stream.
+//  * The caller sends a number of WriteBlobRequests, providing the blob data
+//    in an ordered sequence of chunks.
+//  * The request size should be less than 4 MiB (2^20 = 4,194,304), because
+//    that is the default request limit for gRPC.  Accounting for the other
+//    fields of the request, then, and leaving some padding for safety, the data
+//    chunk should be somewhat smaller-- perhaps simply 4e6 bytes.
+//  * The first of these requests must provide write_blob_first_request_data,
+//    and the subsequent ones must not.
+//  * The last of these requests must set finalize_object=True, and the prior
+//    ones must not.
+//  * For a blob <= 4MB, a single chunk suffices, so the single request is both
+//    the first and the last.
+//  * The caller receives a sequence of WriteBlobResponses in return.  Each
+//    WriteBlobResponse corresponds to one WriteBlobRequest in order, reporting
+//    the state of the write operation.
+//  * The caller may choose to verify each response before sending the next
+//    request, or it may continue sending requests optimistically.
+//
+// Note that the WriteBlobRequest does not mirror the nested structure of
+// WriteScalarRequest, because we only ever send one blob at a time.
+
+// Metadata included with only the first chunk of the blob write stream.
+message WriteBlobFirstRequestData {
+  string experiment_id = 1;
+  string run = 2;
+  string tag = 3;
+  int64 step = 4;
+  int64 index = 5;
+  // Note that metadata.plugin_data.content does not carry the payload.
+  .tensorboard.SummaryMetadata metadata = 6;
+}
+
+// A single chunk of the blob write stream.
+message WriteBlobRequest {
+  WriteBlobFirstRequestData write_blob_first_request_data = 1;
+  bytes data = 2 [ctype = CORD];
+  // The position in the blob where this chunk begins.
+  // This must equal the sum of the sizes of the chunks sent so far.
+  int64 offset = 3;
+  // CRC32C of current data buffer. While optional it is recommended that
+  // clients include the crc32c for every data buffer, to protect against data
+  // corruption. Note that for multi-shot writes, specifying the crc32c for
```

Comment:
We have the flexibility to make these required if you want, even if
they’re not required in the APIs from which these are inspired. What do
you think?

Reply:
I don't feel strongly either way, but sure, may as well.

Reply:
Done (here, just as a matter of documentation).